File: packages/domain/src/validation.ts
import { z } from "zod";

import { type DomainErrorCode } from "./errors.ts";
import { I64_MAX, U32_MAX, U64_MAX } from "./math.ts";
import type { ProtocolConfig } from "./types.ts";

/** The protocol limits that validation consults. A subset of `ProtocolConfig`. */
export type ProtocolLimits = Pick<
  ProtocolConfig,
  | "minBudgetRaw"
  | "maxBudgetRaw"
  | "minEpochSeconds"
  | "maxEpochSeconds"
  | "maxDurationSeconds"
  | "maxEpochs"
  | "maxSpreadBps"
  | "maxDepthBandBps"
  | "minProbeQuoteRaw"
  | "maxProbeQuoteRaw"
  | "minStartLeadSeconds"
  | "positionLockBufferSeconds"
  | "minSetupWindowSeconds"
>;

export interface CreateMandateParams {
  readonly maxRewardRaw: bigint;
  readonly biddingEndsAt: bigint;
  readonly startAt: bigint;
  readonly durationSeconds: bigint;
  readonly epochSeconds: bigint;
  readonly maxEffectiveSpreadBps: number;
  readonly depthBandBps: number;
  readonly minPoolBuyDepthQuoteRaw: bigint;
  readonly minPoolSellDepthQuoteRaw: bigint;
  readonly minProviderQuoteInBandRaw: bigint;
  readonly minProviderBaseQuoteEqInBandRaw: bigint;
  readonly probeQuoteRaw: bigint;
}

/**
 * Validate mandate creation parameters. Returns every distinct problem, sorted; an empty list means valid.
 * The rules and their grouping are the contract shared with the Rust program (golden vectors assert them).
 * Mandate creation is never allowed to succeed on a schedule the reward split cannot honour: the escrow
 * must cover at least one raw unit per epoch.
 */
export function validateCreateMandate(
  params: CreateMandateParams,
  protocol: ProtocolLimits,
  now: bigint,
): DomainErrorCode[] {
  const errors = new Set<DomainErrorCode>();
  const { epochSeconds: e, durationSeconds: d, startAt } = params;

  let totalEpochs: bigint | undefined;
  if (e < protocol.minEpochSeconds || e > protocol.maxEpochSeconds) {
    errors.add("InvalidEpochLength");
  } else if (d <= 0n || d > protocol.maxDurationSeconds) {
    errors.add("InvalidTiming");
  } else if (d % e !== 0n) {
    errors.add("InvalidEpochLength");
  } else {
    totalEpochs = d / e;
    if (totalEpochs > BigInt(protocol.maxEpochs)) errors.add("TooManyEpochs");
  }

  const lockAt = startAt - protocol.positionLockBufferSeconds;
  if (params.biddingEndsAt <= now) errors.add("InvalidTiming");
  if (startAt < now + protocol.minStartLeadSeconds) errors.add("InvalidTiming");
  if (params.biddingEndsAt >= lockAt - protocol.minSetupWindowSeconds) errors.add("InvalidTiming");
  if (
    totalEpochs !== undefined &&
    startAt > 0n &&
    totalEpochs <= BigInt(protocol.maxEpochs) &&
    startAt + e * totalEpochs > I64_MAX
  ) {
    errors.add("ArithmeticOverflow");
  }

  const budget = params.maxRewardRaw;
  if (
    budget < protocol.minBudgetRaw ||
    budget > protocol.maxBudgetRaw ||
    (totalEpochs !== undefined && budget < totalEpochs)
  ) {
    errors.add("InvalidBudget");
  }

  const spread = BigInt(params.maxEffectiveSpreadBps);
  if (spread < 1n || spread > BigInt(protocol.maxSpreadBps)) errors.add("InvalidThreshold");
  const band = BigInt(params.depthBandBps);
  if (band < 1n || band > BigInt(protocol.maxDepthBandBps)) errors.add("InvalidThreshold");
  for (const minimum of [
    params.minPoolBuyDepthQuoteRaw,
    params.minPoolSellDepthQuoteRaw,
    params.minProviderQuoteInBandRaw,
    params.minProviderBaseQuoteEqInBandRaw,
  ]) {
    if (minimum <= 0n) errors.add("InvalidThreshold");
  }
  if (
    params.probeQuoteRaw < protocol.minProbeQuoteRaw ||
    params.probeQuoteRaw > protocol.maxProbeQuoteRaw
  ) {
    errors.add("InvalidThreshold");
  }
  return [...errors].sort();
}

export interface BidParams {
  readonly requestedRewardRaw: bigint;
  readonly validUntil: bigint;
  readonly maxRewardRaw: bigint;
  readonly now: bigint;
  readonly totalEpochs: number;
  /** Last instant a sponsor can still accept: the start of the provider's setup window. */
  readonly acceptanceCutoff: bigint;
}

/** Validate a bid. Sorted distinct problems; empty means valid. */
export function validateBid(bid: BidParams): DomainErrorCode[] {
  const errors = new Set<DomainErrorCode>();
  if (
    bid.requestedRewardRaw <= 0n ||
    bid.requestedRewardRaw > bid.maxRewardRaw ||
    bid.requestedRewardRaw < BigInt(bid.totalEpochs)
  ) {
    errors.add("InvalidBudget");
  }
  if (bid.validUntil < bid.now) errors.add("BidExpired");
  else if (bid.validUntil > bid.acceptanceCutoff) errors.add("InvalidTiming");
  return [...errors].sort();
}

// ---- JSON transport schemas -------------------------------------------------------------------
// HTTP/JSON carries u64 and i64 as base-10 strings. These schemas parse them into bigint and
// enforce the fixed-width ranges, so nothing downstream ever sees a JavaScript number as money.

const uintString = z
  .string()
  .regex(/^(0|[1-9][0-9]*)$/, "expected a base-10 unsigned integer string")
  .transform((text) => BigInt(text));

export const u64String = uintString.refine((x) => x <= U64_MAX, "exceeds u64");
export const i64String = z
  .string()
  .regex(/^(0|[1-9][0-9]*)$/, "expected a base-10 integer string")
  .transform((text) => BigInt(text))
  .refine((x) => x <= I64_MAX, "exceeds i64");
const u32Number = z.number().int().min(0).max(Number(U32_MAX));

export const createMandateParamsSchema = z.strictObject({
  maxRewardRaw: u64String,
  biddingEndsAt: i64String,
  startAt: i64String,
  durationSeconds: i64String,
  epochSeconds: i64String,
  maxEffectiveSpreadBps: u32Number,
  depthBandBps: u32Number,
  minPoolBuyDepthQuoteRaw: u64String,
  minPoolSellDepthQuoteRaw: u64String,
  minProviderQuoteInBandRaw: u64String,
  minProviderBaseQuoteEqInBandRaw: u64String,
  probeQuoteRaw: u64String,
});

export const bidParamsSchema = z.strictObject({
  requestedRewardRaw: u64String,
  validUntil: i64String,
});

File: crates/mandate-core/src/validation.rs
//! Validation of mandate creation and bid parameters. Returns every distinct problem so callers can
//! explain all of them at once. The rules, their grouping and the resulting error names are the
//! contract shared with `@mandate/domain` and asserted by the golden vectors.
//!
//! All comparisons widen to `i128`, so no combination of inputs can overflow.
use crate::MandateCoreError;

/// The protocol limits validation consults.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ProtocolLimits {
    pub min_budget_raw: u64,
    pub max_budget_raw: u64,
    pub min_epoch_seconds: i64,
    pub max_epoch_seconds: i64,
    pub max_duration_seconds: i64,
    pub max_epochs: u32,
    pub max_spread_bps: u32,
    pub max_depth_band_bps: u32,
    pub min_probe_quote_raw: u64,
    pub max_probe_quote_raw: u64,
    pub min_start_lead_seconds: i64,
    pub position_lock_buffer_seconds: i64,
    pub min_setup_window_seconds: i64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CreateMandateParams {
    pub max_reward_raw: u64,
    pub bidding_ends_at: i64,
    pub start_at: i64,
    pub duration_seconds: i64,
    pub epoch_seconds: i64,
    pub max_effective_spread_bps: u32,
    pub depth_band_bps: u32,
    pub min_pool_buy_depth_quote_raw: u64,
    pub min_pool_sell_depth_quote_raw: u64,
    pub min_provider_quote_in_band_raw: u64,
    pub min_provider_base_quote_eq_in_band_raw: u64,
    pub probe_quote_raw: u64,
}

/// A set of distinct error codes.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct ErrorSet(u32);

impl ErrorSet {
    pub fn insert(&mut self, error: MandateCoreError) {
        self.0 |= Self::bit(error);
    }

    pub const fn contains(&self, error: MandateCoreError) -> bool {
        self.0 & Self::bit(error) != 0
    }

    pub const fn is_empty(&self) -> bool {
        self.0 == 0
    }

    const fn bit(error: MandateCoreError) -> u32 {
        match 1u32.checked_shl(error as u32) {
            Some(bit) => bit,
            None => 0,
        }
    }

    /// The highest-priority error (declaration order), for callers that can return only one.
    pub fn first(&self) -> Option<MandateCoreError> {
        self.iter().next()
    }

    /// Members in declaration order.
    pub fn iter(&self) -> impl Iterator<Item = MandateCoreError> + '_ {
        MandateCoreError::ALL
            .into_iter()
            .filter(|error| self.contains(*error))
    }
}

/// Validate mandate creation parameters. Escrow must cover at least one raw unit per epoch.
pub fn validate_create_mandate(
    params: &CreateMandateParams,
    protocol: &ProtocolLimits,
    now: i64,
) -> ErrorSet {
    let mut errors = ErrorSet::default();
    let epoch = i128::from(params.epoch_seconds);
    let duration = i128::from(params.duration_seconds);
    let start = i128::from(params.start_at);

    let mut total_epochs: Option<i128> = None;
    if params.epoch_seconds < protocol.min_epoch_seconds
        || params.epoch_seconds > protocol.max_epoch_seconds
    {
        errors.insert(MandateCoreError::InvalidEpochLength);
    } else if params.duration_seconds <= 0
        || params.duration_seconds > protocol.max_duration_seconds
    {
        errors.insert(MandateCoreError::InvalidTiming);
    } else if let (Some(rem), Some(total)) =
        (duration.checked_rem(epoch), duration.checked_div(epoch))
    {
        if rem != 0 {
            errors.insert(MandateCoreError::InvalidEpochLength);
        } else {
            total_epochs = Some(total);
            if total > i128::from(protocol.max_epochs) {
                errors.insert(MandateCoreError::TooManyEpochs);
            }
        }
    } else {
        // Only reachable if protocol.min_epoch_seconds <= 0: a zero epoch is never valid.
        errors.insert(MandateCoreError::InvalidEpochLength);
    }

    let lock_at = start.saturating_sub(i128::from(protocol.position_lock_buffer_seconds));
    if params.bidding_ends_at <= now {
        errors.insert(MandateCoreError::InvalidTiming);
    }
    if start < i128::from(now).saturating_add(i128::from(protocol.min_start_lead_seconds)) {
        errors.insert(MandateCoreError::InvalidTiming);
    }
    if i128::from(params.bidding_ends_at)
        >= lock_at.saturating_sub(i128::from(protocol.min_setup_window_seconds))
    {
        errors.insert(MandateCoreError::InvalidTiming);
    }
    if let Some(total) = total_epochs {
        let end = epoch
            .checked_mul(total)
            .and_then(|span| start.checked_add(span));
        if start > 0
            && total <= i128::from(protocol.max_epochs)
            && end.is_none_or(|value| value > i128::from(i64::MAX))
        {
            errors.insert(MandateCoreError::ArithmeticOverflow);
        }
    }

    let budget = params.max_reward_raw;
    let below_epoch_count = total_epochs.is_some_and(|total| i128::from(budget) < total);
    if budget < protocol.min_budget_raw || budget > protocol.max_budget_raw || below_epoch_count {
        errors.insert(MandateCoreError::InvalidBudget);
    }

    if params.max_effective_spread_bps < 1
        || params.max_effective_spread_bps > protocol.max_spread_bps
        || params.depth_band_bps < 1
        || params.depth_band_bps > protocol.max_depth_band_bps
        || params.min_pool_buy_depth_quote_raw == 0
        || params.min_pool_sell_depth_quote_raw == 0
        || params.min_provider_quote_in_band_raw == 0
        || params.min_provider_base_quote_eq_in_band_raw == 0
        || params.probe_quote_raw < protocol.min_probe_quote_raw
        || params.probe_quote_raw > protocol.max_probe_quote_raw
    {
        errors.insert(MandateCoreError::InvalidThreshold);
    }
    errors
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct BidParams {
    pub requested_reward_raw: u64,
    pub valid_until: i64,
    pub max_reward_raw: u64,
    pub now: i64,
    pub total_epochs: u32,
    /// Last instant a sponsor can still accept: the start of the provider's setup window.
    pub acceptance_cutoff: i64,
}

/// Validate a bid.
pub fn validate_bid(bid: &BidParams) -> ErrorSet {
    let mut errors = ErrorSet::default();
    if bid.requested_reward_raw == 0
        || bid.requested_reward_raw > bid.max_reward_raw
        || bid.requested_reward_raw < u64::from(bid.total_epochs)
    {
        errors.insert(MandateCoreError::InvalidBudget);
    }
    if bid.valid_until < bid.now {
        errors.insert(MandateCoreError::BidExpired);
    } else if bid.valid_until > bid.acceptance_cutoff {
        errors.insert(MandateCoreError::InvalidTiming);
    }
    errors
}
